cleanup of cs_modify.cpp

KarrokDC opened the pull request:
**Changes proposed**:
- Remove a bunch of duplicate code

**Target branch(es)**: 335/6x

**Issues addressed**: Fixes

**Tests performed**: (Does it build, tested in-game, etc)
- [x] Build
- [x] Run

Commands tested for expected results (ie changes on valid input, error message on invalid input):
- [x] .modify hp 15000
- [x] .modify hp -1000 
- [x] .modify hp aaa
- [x] .modify energy 100
- [x] .modify energy -10
- [x] .modify energy aaa
- [x] .modify money 1000
- [x] .modify money 1000g
- [x] .modify money -1000g
- [x] .modify money -10000000g
- [x] .modify scale 2
- [x] .modify scale 0.5
- [x] .modify scale -1

Commands tested on other player:
- [x] Same list as above but on another player

**Known issues and TODO list**:

A number of notification messages still have a "wrong" order compared to the new NotifyModification which uses Namelink , value in both the regular handle and in the reporting notification. Whereas (for example)HandleModifyGenderCommand and HandleModifyMoneyCommand still use inconsistent order versions of notifications. @Nyeriah Fixed a couple of these order issues in https://github.com/TrinityCore/TrinityCore/commit/c1905771f064e374a09720f470116f04b03e5a42

Modifying powers (energy) currently has a awkward *10 modifier. This can potentially be removed completely as it also gives inconsistent results. ie. .modify energy 200 actually set your rogue's energy to 2000. For whatever reason, this does not apply for rage and runicpower even though these also have the initial x10 multiplier.

Furthermore this PR is based on @chaodhib  refactoring of SetSpeed: https://github.com/TrinityCore/TrinityCore/pull/16843/commits/4f0973bd3534839caf7b834958775acaff963fd9
So his PR should probably be merged in before this one is considered.

I don't know what has to be done to change the parameter order as @Nyeriah has done, so perhaps she could look into that?

DDuarte commented:
Note to self: this PR needs #16843 to be merged first, only the changes from https://github.com/TrinityCore/TrinityCore/pull/16858/commits/23c2a37be65b88de52a2c33eca4a4831a44b8123 are to be considered for this PR.

Ascathor commented:
I like what I see. *nods*

tkrokli commented:
Agreed. It is nice to see code cleanup and simplification like this. :+1:

DDuarte commented:
@KarrokDC: #16843 has been merged, can you rebase this PR with your changes only?

KarrokDC committed:
Removed a chunk of duplicate code

Generalized Message notification
Generalized checking of resources modification with a min-max
Generalized checking of single value modification

KarrokDC commented:
@DDuarte done.

Naios commented on src/server/scripts/Commands/cs_modify.cpp:
```diff
@@ -80,23 +80,32 @@ class modify_commandscript : public CommandScript
-    //Edit Player HP
-    static bool HandleModifyHPCommand(ChatHandler* handler, const char* args)
+    template<typename... Args>
+    static void NotifyModification(ChatHandler* handler, Unit* target, const TrinityStrings resourceMessage, const TrinityStrings resourceReportMessage, Args&&... args)
```
Const non reference qualified parameters are converted into it's decayed version (the constness is dropped internally), which makes it unneccesarry to add const to the params here.

Naios commented on src/server/scripts/Commands/cs_modify.cpp:
```diff
@@ -80,23 +80,32 @@ class modify_commandscript : public CommandScript
+        {
+            handler->PSendSysMessage(resourceMessage, handler->GetNameLink(player).c_str(), std::forward<Args>(args)...);
+            if (handler->needReportToTarget(player))
+                ChatHandler(player->GetSession()).PSendSysMessage(resourceReportMessage, handler->GetNameLink().c_str(), std::forward<Args>(args)...);
```
Never forward a collapsing reference twice! For example forwarding a r-value qualified string two times will move the string the first time which means the string will be empty the second time you forward it.

Naios commented on src/server/scripts/Commands/cs_modify.cpp:
```diff
@@ -107,164 +116,87 @@ class modify_commandscript : public CommandScript
-        if (mana <= 0 || manam <= 0 || manam < mana)
+        int32 hp, hpmax;
+        Player* target = handler->getSelectedPlayerOrSelf();
+        if (CheckModifyResources(handler, args, target, hp, hpmax))
```
A good principle in software engineering is to terminate control flows early, which means:

```c++
if (!CheckModifyResources(handler, args, target, hp, hpmax))
    return false;
```

is preferrable over

```c++
if (CheckModifyResources(handler, args, target, hp, hpmax))
{
    ...
    return true;
}

return false;
```

You don't need to change this, I just mentioned it because it might help you in the future.

Naios commented on src/server/scripts/Commands/cs_modify.cpp:
```diff
@@ -107,164 +116,87 @@ class modify_commandscript : public CommandScript
+        int32 energy, energymax;
         Player* target = handler->getSelectedPlayerOrSelf();
-        if (!target)
+        const int8 energyMultiplier = 10;
```
TC uses the `[type] const [identifier]` codestyle

Naios commented:
Merged in 775b03bf6f, thanks for your great PR!

The cherry-pick to 6.x will take some time because I need to resolve the conflicts with the commits related to #16648 (#16751 and c1905771f064e37) first.

Naios commented:
The commit was picked to 6.x in 8f05a4779313b1f

KarrokDC commented:
@Naios Thanks for the review and comments! :)
